File: leetcode-in-python3/solution/122_best_time_to_buy_and_sell_stock_ii.py
from typing import List
import unittest


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        ans = 0
        for i, _ in enumerate(prices[1:],1):
            if prices[i-1] < prices[i]:
                ans += prices[i] - prices[i-1]
        return ans


class TestSolution(unittest.TestCase):

    def setUp(self):
        self.test_case = [
            ([7, 1, 5, 3, 6, 4], 7),
            ([7, 6, 4, 3, 1], 0),
            ([5], 0),
            ([4, 5], 1),
        ]
        self.s = Solution()

    def test_solution(self):
        for nums, answer in self.test_case:
            ans = self.s.maxProfit(nums)
            self.assertEqual(answer,ans , (nums, answer))


if __name__ == '__main__':
    unittest.main()

File: leetcode-in-rust/src/solution/_122_best_time_to_buy_and_sell_stock_ii.rs
pub struct Solution {}

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut max_profit = 0;
        let mut pre_price = prices[0];
        for cur_price in prices{
            if pre_price < cur_price{
                max_profit += cur_price - pre_price;
            }
            pre_price = cur_price;
        }
        max_profit
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solution() {
        let test_case = vec![
            (vec![7, 1, 5, 3, 6, 4], 7),
            (vec![7, 6, 4, 3, 1], 0),
            (vec![5], 0),
            (vec![4, 5], 1),
        ];
        for (nums, answer) in test_case {
            let ans = Solution::max_profit(nums);
            assert_eq!(ans, answer);
        }
    }
}
